chain/types/ethtypes/eth_legacy_155_transactions.go

```diff
@@ -0,0 +1,289 @@
+	if vValue.Equals(big.NewInt(27)) {
+		sig[64] = 0
+	} else if vValue.Equals(big.NewInt(28)) {
+		sig[64] = 1
+	} else {
+		return nil, fmt.Errorf("invalid 'v' value: expected 27 or 28, got %d", vValue.Int64())
+	}
+
+	return sig[0:65], nil
+}
+
+func (tx *EthLegacy155TxArgs) InitialiseSignature(sig typescrypto.Signature) error {
+	if sig.Type != typescrypto.SigTypeDelegated {
+		return fmt.Errorf("RecoverSignature only supports Delegated signature")
+	}
+
+	if len(sig.Data) != EthLegacy155TxSignatureLen {
+		return fmt.Errorf("signature should be %d bytes long, but got %d bytes", EthLegacy155TxSignatureLen, len(sig.Data))
+	}
+
+	if sig.Data[0] != EthLegacy155TxSignaturePrefix {
+		return fmt.Errorf("expected signature prefix 0x01, but got 0x%x", sig.Data[0])
+	}
+
+	// ignore the first byte of the tx as it's only used for legacy transaction identification
+	r_, err := parseBigInt(sig.Data[1:33])
+	if err != nil {
+		return fmt.Errorf("cannot parse r into EthBigInt: %w", err)
+	}
+
+	s_, err := parseBigInt(sig.Data[33:65])
+	if err != nil {
+		return fmt.Errorf("cannot parse s into EthBigInt: %w", err)
+	}
+
+	v_, err := parseBigInt(sig.Data[65:])
+	if err != nil {
+		return fmt.Errorf("cannot parse v into EthBigInt: %w", err)
+	}
+
+	if err := validateEIP155ChainId(v_); err != nil {
+		return fmt.Errorf("failed to validate EIP155 chain id: %w", err)
+	}
+
+	tx.LegacyTx.R = r_
+	tx.LegacyTx.S = s_
+	tx.LegacyTx.V = v_
+	return nil
+}
+
+func (tx *EthLegacy155TxArgs) packTxFields() ([]interface{}, error) {
+	nonce, err := formatInt(tx.LegacyTx.Nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	// format gas price
+	gasPrice, err := formatBigInt(tx.LegacyTx.GasPrice)
+	if err != nil {
+		return nil, err
+	}
+
+	gasLimit, err := formatInt(tx.LegacyTx.GasLimit)
+	if err != nil {
+		return nil, err
+	}
+
+	value, err := formatBigInt(tx.LegacyTx.Value)
+	if err != nil {
+		return nil, err
+	}
+
+	chainIdBigInt := big.NewIntUnsigned(build.Eip155ChainId)
+	chainId, err := formatBigInt(chainIdBigInt)
+	if err != nil {
+		return nil, err
+	}
+
+	r, err := formatInt(0)
+	if err != nil {
+		return nil, err
+	}
+
+	s, err := formatInt(0)
+	if err != nil {
+		return nil, err
+	}
+
+	res := []interface{}{
+		nonce,
+		gasPrice,
+		gasLimit,
+		formatEthAddr(tx.LegacyTx.To),
+		value,
+		tx.LegacyTx.Input,
+		chainId,
+		r, s,
+	}
+	return res, nil
+}
+
+func validateEIP155ChainId(v big.Int) error {
+	chainId := deriveEIP155ChainId(v)
+	if !chainId.Equals(big.NewIntUnsigned(build.Eip155ChainId)) {
+		return fmt.Errorf("invalid chain id, expected %d, got %s", build.Eip155ChainId, chainId.String())
+	}
+	return nil
+}
+
+// deriveEIP155ChainId derives the chain id from the given v parameter
+func deriveEIP155ChainId(v big.Int) big.Int {
+	if big.BitLen(v) <= 64 {
+		vUint64 := v.Uint64()
+		if vUint64 == 27 || vUint64 == 28 {
+			return big.NewInt(0)
+		}
+		return big.NewIntUnsigned((vUint64 - 35) / 2)
```

Comment:
@ZenGround0 Please can you elaborate on the question ? I don't understand this one.

Reply:
Which other valid value ?

Reply:
I meant other valid value: `big.NewIntUnsigned((vUint64 - 36) / 2)`.  Depending on the value of r and the curve when creating the signature the message creator could have set `v = chainID * 2 +36`

Reply:
Actually now I'm confused how you can deterministically recover without more information.  Take for example a chain ID of 22.  If parity is 0 then v value is 79.  But for a chain id of 21 and a parity of 1 we also encode a v value of 79
```
 (79 - 35)/2: 22
 (79 - 36)/2: 21
```
So I think you'll need to determine the correct parity from the r value before decoding .

Reply:
It would be worth seeing how ethereum does this.  Maybe there are some conventions around always choosing one parity or the other.

Reply:
@ZenGround0 Posted on Filecoin Slack but reproducing here for visibility:

So I dug deep into this.

- Looks like go-ethereum  simply assumes a "recovery_id"(AKA "Y parity") of 0 when extracting the ChainId  from the V  value:
   - https://github.com/ethereum/go-ethereum/blob/804afb8faa0e3796fb8e6f39beabce356bc98298/core/types/transaction_signing.go#L567
    - https://github.com/ethereum/go-ethereum/blob/804afb8faa0e3796fb8e6f39beabce356bc98298/core/types/transaction.go#L225
  - There is no code in go-ethereum  that somehow manipulates this V  value to detect the "recovery_id" before deriving the ChainID  from it

**_Having said that, this is not a big concern because:_**
The only check we need to perform is that the user has signed the correct ChainId  and that is done here by inserting the expected ChainID  in the digest:
https://github.com/filecoin-project/lotus/blob/0fd4ede7e7a0637cf639e29bd694c7c686b7abe7/chain/types/ethtypes/eth_legacy_155_transactions.go#L232

- Really, this is all we need. Once past this, we don't need to validate the ChainId /derive the ChainID  from the V  value. We only really do it as a sanity check.
- So we can just completely get rid of deriveEIP155ChainId /validateEIP155ChainId call after ensuring that the user has signed the correct digest
- I am however still inclined to keep this around because in all my testing with CoinBase Wallet etc,  the correct ChainId  is indeed derived
- If at all in the future, users complain of their EIP-155 transactions not going through,  we can remove this check/look into it

Let me know what you think and thanks for eagle eyes :eyes:
I also agree that we should get Raul/Steb/someone else with deep ETH knowledge to look at this before we ship it but want to make sure you are satisfied here before we move on to that. Thanks a lot. (edited)

Reply:
This has been resolved offline with @ZenGround0